flang/lib/Lower/PFTBuilder.cpp

```diff
@@ -191,16 +191,35 @@ class PFTBuilder {
     resetFunctionState();
   }
 
-  /// Ensure that a function ends with a valid branch target (and is nonempty).
+  /// Add the end statement Evaluation of a sub/program to the PFT.
+  /// There may be intervening internal subprogram definitions between
+  /// prior statements and this end statement.
   void endFunctionBody() {
     if (evaluationListStack.empty())
       return;
     auto evaluationList = evaluationListStack.back();
-    if (evaluationList->empty() ||
-        !evaluationList->back().isA<parser::ContinueStmt>()) {
-      static const parser::ContinueStmt endTarget{};
-      addEvaluation(
-          lower::pft::Evaluation{endTarget, parentVariantStack.back(), {}, {}});
+    if (evaluationList->empty() || !evaluationList->back().isEndStmt()) {
+      const auto &endStmt =
+          parentVariantStack.back().get<lower::pft::FunctionLikeUnit>().endStmt;
+      endStmt.visit(common::visitors{
+          [&](const parser::Statement<parser::EndProgramStmt> &s) {
+            addEvaluation(lower::pft::Evaluation{
+                s.statement, parentVariantStack.back(), s.source, s.label});
+          },
+          [&](const parser::Statement<parser::EndFunctionStmt> &s) {
+            addEvaluation(lower::pft::Evaluation{
+                s.statement, parentVariantStack.back(), s.source, s.label});
+          },
+          [&](const parser::Statement<parser::EndSubroutineStmt> &s) {
+            addEvaluation(lower::pft::Evaluation{
+                s.statement, parentVariantStack.back(), s.source, s.label});
+          },
+          [&](const parser::Statement<parser::EndMpSubprogramStmt> &s) {
+            addEvaluation(lower::pft::Evaluation{
+                s.statement, parentVariantStack.back(), s.source, s.label});
+          },
+          [&](auto &) {},
```

Comment:
You could probably add a static_assert here that this is not an end statements so that we do not forgot to update this if new end statement are added:
```
[&](const auto &s) {
 static_assert(!lower::pft::isEndStmt<std::decay_t<decltype(s)>, "missing handling for end statements");
},
```

Reply:
That's a good idea, but it doesn't quite work.  The 4 existing end statements in `ReferenceVariant`  `FunctionStatement` each have a corresponding begin statement.  Those begin statements do not have Evaluation's, so isEndStmt and any other reference that assumes Evaluation status breaks the build.  An attempt to use an obvious template here doesn't work for the same reason.  A runtime `llvm_unreachable` doesn't have that problem, so I've added that.

Reply:
Ah, I just missed this for this merge.  We can consider this for a future change.  Thanks for the suggestion.

Reply:
Right, I see why the static assert is bound to fail here, sorry for the bad hint.